src/runtime_src/core/common/xrt_xclbin.cpp

```diff
@@ -74,3 +78,348 @@ xrtXclbinUUID(xclDeviceHandle dhdl, xuid_t out)
     return -1;
   }
 }
+
+namespace xrt {
+
+// class xclbin_impl - class for xclbin objects
+//
+// Life time of xclbin are managed through shared pointers.
+// A buffer is freed when last references is released.
+class xclbin_impl
+{
+protected:
+  std::vector<char> m_axlf;
+  const axlf* m_top;
+
+  void init_axlf_handle()
+  {
+    const axlf* tmp = reinterpret_cast<const axlf*>(m_axlf.data());
+    if (strncmp(tmp->m_magic, "xclbin2", 7)) // Future: Do not hardcode "xclbin2"
+      throw std::runtime_error("Invalid xclbin");
+    m_top = tmp;
+  }
+
+public:
+  xclbin_impl(const std::vector<char>& data)
+  {
+    m_axlf = data;
+    // Set pointer of type axlf*, and check magic string
+    init_axlf_handle();
+  }
+
+  xclbin_impl(const std::string& filename)
+  {
+    if (filename.empty())
+      throw std::runtime_error("No XCLBIN specified");
+
+    // Load the file into data
+    std::ifstream stream(filename);
+    stream.seekg(0, stream.end);
+    size_t size = stream.tellg();
+    stream.seekg(0, stream.beg);
+    m_axlf.resize(size);
+    stream.read(m_axlf.data(), size);
+
+    // Set pointer of type axlf*, and check magic string
+    init_axlf_handle();
+  }
+
+  std::vector<std::string>
+  get_cu_names() const
+  {
+    std::vector<std::string> names;
+    for (auto cu : xrt_core::xclbin::get_cus(m_top, false)) // Why would I ever need them encoded
+      names.push_back(xrt_core::xclbin::get_ip_name(m_top, cu));
+
+    return names;
+  }
+
+  // TODO - Check that m_header is present in data buffer
+  std::string
+  get_xsa_name() const
+  {
+    return reinterpret_cast<const char*>(m_top->m_header.m_platformVBNV);
+  }
+
+  // TODO - Check that m_header is present in data buffer
+  uuid
+  get_uuid() const
+  {
+    return {m_top->m_header.uuid};
```

Comment:
I don't believe curly braces are needed.

Reply:
I believe it was recommend to ensure safe copy/conversion. You are correct the curly braces are not needed. I verified the code, it is safe to remove.